### PR TITLE
Validate OCF date boundaries

### PR DESCRIPTION
**Pull request by HardlyDifficult**
## Summary
- replace all raw timestamp splitting with one validated date boundary
- accept canonical YYYY-MM-DD and RFC 3339 timestamps with Z or numeric offsets
- reject wrong runtime types, malformed timestamps, impossible calendar dates, junk suffixes, and year zero
- preserve lexical OCF date semantics across timezone offsets
- add a non-throwing date parser for comparison code and harden date-to-DAML conversion through the same validator
- attach contextual field paths at converter boundaries
- require discriminator-correlated convertible and warrant triggers with caller- or ledger-owned stable IDs
- expose schema-exact stock-class ratio conversion rights while keeping the DAML-v34 storage sentinel private
- reject targetless writes and any generated decode/encode path that would discard or alter DAML data

## Validation
- typecheck, clean-state lint, formatting, build, declarations, and diff checks
- 61 unit suites / 3,033 tests passing
- focused date, trigger, vesting, transaction-ordering, and stock-class conversion-right regressions
- property tests and nested converter regressions
- zero remaining raw split-on-T date conversions in source
- exact-head GitHub build-and-test and 8m22s OCP QuickStart are green; Copilot completed with no findings; Cursor finding was resolved with generated-code and ledger-wire evidence; all substantive CodeRabbit findings are addressed on `02bb2a1`; exact-head Copilot, Cursor, CodeRabbit, build, and QuickStart are being re-run

## Stack
- based on https://github.com/Fairmint/ocp-canton-sdk/pull/410
- depends transitively on https://github.com/Fairmint/ocp-canton-sdk/pull/409 and https://github.com/Fairmint/ocp-canton-sdk/pull/408

---

> [!NOTE]
> **Medium Risk**
> Touches many cap-table transaction converters and stock-class conversion semantics; behavior changes on invalid dates, triggers, or non-ratio stock-class rights could reject previously tolerated ledger or input shapes.
> 
> **Overview**
> **Date handling** is centralized: converters stop using raw `split('T')` and instead route through `damlTimeToDateString`, `dateStringToDAMLTime`, and optional/nullable variants, each taking a **field path** so validation errors name the offending OCF field.
> 
> **Convertible issuance** gains stricter trigger modeling: shared `triggerFieldsToDaml` / `triggerFieldsFromDaml` enforce discriminator-specific trigger shapes, require stable `trigger_id`s (no synthetic fallbacks on read), and tighten conversion-right and note-interest-rate validation with indexed paths. **Vesting** rows use shared `filterAndMapVestingsToDaml` with non-negative amount checks.
> 
> **Stock class conversion rights** are narrowed to canonical OCF **ratio** mechanisms only; DAML v34’s required but unused `conversion_trigger` is written and read via a **private storage sentinel** (fixed trigger id + custom mechanism description), with decode rejecting unsupported DAML-only fields and **lossy** generated decode/encode round-trips. `damlStockClassDataToNative` now decodes through the generated DAML codec with those guards.
> 
> Similar **path-aware** date and collection validation is applied across equity compensation, plan security, issuer/stock-class adjustments, and other OpenCapTable transaction converters.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 370f61acdccc7e3ecb9f68a229b11bcb6cfe3a17. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added shared trigger-field conversion with discriminator-aware validation and consistent DAML↔OCF date serialization.

* **Bug Fixes**
  * Standardized ISO/RFC3339 date parsing to canonical `YYYY-MM-DD` output, fixing timezone-offset and impossible-date edge cases.
  * Improved optional vs nullable date handling (including omission behavior) and strengthened index-aware, field-path-specific validation errors across conversions.

* **Tests**
  * Expanded date-boundary, trigger-field, schema/date-invariant, and property-based conversion test coverage, plus added regression cases for precise error diagnostics.

### CODE REVIEW

File: src/utils/typeConversions.ts [hunk 1]
```diff
@@ -24,17 +24,65 @@ export function isRecord(value: unknown): value is Record<string, unknown> {
...
+    code: OcpErrorCodes.INVALID_FORMAT,
+  });
+}
+
+/**
+ * Convert a valid OCF date to DAML Time format. RFC 3339 date-times are validated and preserved exactly; date-only
+ * values receive a UTC midnight suffix.
+ */
+export function dateStringToDAMLTime(value: string, fieldPath?: string): string;
+export function dateStringToDAMLTime(value: unknown, fieldPath = 'date'): string {
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/utils/typeConversions.ts]: Fixed in `d7ef4f1`: removed the string-only overload so the exported signature now accepts `unknown`, matching the runtime validator. Unit and property tests call non-string values directly (no Reflect/casts), and a built-declaration probe passes an `unknown` value through the published utils entry point. Targeted 107 tests, build, declarations, typecheck, all 58 suites / 2,734 tests, lint, Prettier, and diff check pass.

File: src/functions/OpenCapTable/convertibleIssuance/getConvertibleIssuanceAsOcf.ts [hunk 2]
```diff
@@ -636,15 +651,22 @@ export function damlConvertibleIssuanceDataToNative(d: Record<string, unknown>):
...
       : {}),
     ...(typeof d.stockholder_approval_date === 'string' && d.stockholder_approval_date.length
-      ? { stockholder_approval_date: d.stockholder_approval_date.split('T')[0] }
+      ? {
+          stockholder_approval_date: damlTimeToDateString(
+            d.stockholder_approval_date,
+            'convertibleIssuance.stockholder_approval_date'
+          ),
+        }
       : {}),
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/convertibleIssuance/getConvertibleIssuanceAsOcf.ts]: Fixed in a21d3fa: convertible issuance approval dates are now omitted only for null/undefined; every present value flows through strict date validation. Both board and stockholder fields have runtime regressions asserting INVALID_TYPE, exact field path, and received value for non-string ledger data. Targeted 42 tests, build, declarations/typecheck, all 58 suites / 2,738 tests, lint, Prettier, and diff check pass.

File: src/functions/OpenCapTable/warrantIssuance/getWarrantIssuanceAsOcf.ts [hunk 3]
```diff
@@ -488,14 +495,26 @@ export function damlWarrantIssuanceDataToNative(d: Record<string, unknown>): Ocf
...
       : {}),
     ...(d.stockholder_approval_date && typeof d.stockholder_approval_date === 'string'
-      ? { stockholder_approval_date: d.stockholder_approval_date.split('T')[0] }
+      ? {
+          stockholder_approval_date: damlTimeToDateString(
+            d.stockholder_approval_date,
+            'warrantIssuance.stockholder_approval_date'
+          ),
+        }
       : {}),
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/warrantIssuance/getWarrantIssuanceAsOcf.ts]: Fixed in a21d3fa: warrant issuance approval dates are now omitted only for null/undefined; every present value flows through strict date validation. Both board and stockholder fields have runtime regressions asserting INVALID_TYPE, exact field path, and received value for non-string ledger data. Targeted 42 tests, build, declarations/typecheck, all 58 suites / 2,738 tests, lint, Prettier, and diff check pass.

File: src/functions/OpenCapTable/issuerAuthorizedSharesAdjustment/getIssuerAuthorizedSharesAdjustmentAsOcf.ts [hunk 4]
```diff
@@ -49,14 +49,26 @@ export function damlIssuerAuthorizedSharesAdjustmentDataToNative(
...
     issuer_id: d.issuer_id,
     new_shares_authorized: normalizeNumericString(
       typeof d.new_shares_authorized === 'number' ? String(d.new_shares_authorized) : d.new_shares_authorized
     ),
-    ...(d.board_approval_date ? { board_approval_date: (d.board_approval_date as string).split('T')[0] } : {}),
+    ...(d.board_approval_date
+      ? {
+          board_approval_date: damlTimeToDateString(
+            d.board_approval_date,
+            'issuerAuthorizedSharesAdjustment.board_approval_date'
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/issuerAuthorizedSharesAdjustment/getIssuerAuthorizedSharesAdjustmentAsOcf.ts]: Fixed in ea9dac1: null/undefined alone now mean absent; every present value passes through the strict date validator. Regression matrices cover empty-string INVALID_FORMAT, non-string INVALID_TYPE, and null handling across the affected read/write boundaries. The same-class convertible trigger, approval, and interest-date guards were hardened too. Local gate passes build, declarations, typecheck, lint, Prettier, 3 focused suites / 100 tests, and all 58 suites / 2,790 tests.

File: src/functions/OpenCapTable/convertibleIssuance/createConvertibleIssuance.ts [hunk 5]
```diff
@@ -447,14 +465,23 @@ export function convertibleIssuanceDataToDaml(d: {
   seniority: number;
   comments?: string[];
 }): Fairmint.OpenCapTable.OCF.ConvertibleIssuance.ConvertibleIssuanceOcfData {
+  const boardApprovalDate: unknown = d.board_approval_date;
+  const stockholderApprovalDate: unknown = d.stockholder_approval_date;
+
   return {
     id: d.id,
     date: dateStringToDAMLTime(d.date),
     security_id: d.security_id,
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/convertibleIssuance/createConvertibleIssuance.ts:520]: Fixed in de8b391: the required write date now calls dateStringToDAMLTime with convertibleIssuance.date, and a regression asserts INVALID_FORMAT, the exact contextual field path, and the received value. Local gate passes build, declarations, typecheck, all 58 suites / 2,791 tests, lint, and Prettier.

File: test/converters/convertibleIssuanceConverters.test.ts [hunk 6]
```diff
@@ -41,6 +42,47 @@ const SAFE_TRIGGER_BASE = {
...
+  code: OcpErrorCode = OcpErrorCodes.INVALID_FORMAT
+): void {
+  try {
+    action();
+    throw new Error('Expected date validation to fail');
+  } catch (error) {
+    expect(error).toBeInstanceOf(OcpValidationError);
+    expect(error).toMatchObject({ code, fieldPath, receivedValue });
+  }
+}
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply test/converters/convertibleIssuanceConverters.test.ts]: Fixed in 16a3232. The duplicate helpers are replaced by a shared single-execution assertion that uses toThrow(OcpValidationError) and preserves code, fieldPath, and receivedValue checks. It is now used by convertible, warrant, and general date-boundary tests. Full local gate: 60 suites / 2,913 tests with coverage, build, declarations, typecheck, zero-warning lint, Prettier, and diff check.

File: test/converters/convertibleIssuanceConverters.test.ts [hunk 7]
```diff
@@ -405,6 +447,283 @@ describe('SAFE conversion_timing round-trip', () => {
...
+        damlConvertibleIssuanceDataToNative({
+          ...BASE_DAML,
+          convertible_type: 'OcfConvertibleNote',
+          conversion_triggers: [trigger],
+        }),
+      `${NOTE_INTEREST_RATE_PATH}.accrual_end_date`,
+      invalidDate,
+      code
+    );
+  });
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply test/converters/convertibleIssuanceConverters.test.ts:779]: Fixed in 16a3232. The accrual_end_date matrix now labels empty and non-string cases in both data and the test title. The same change also adds exact nested trigger/rate index regressions. Full local gate: 60 suites / 2,913 tests with coverage, build, declarations, typecheck, zero-warning lint, Prettier, and diff check.

File: src/functions/OpenCapTable/equityCompensationIssuance/createEquityCompensationIssuance.ts [hunk 8]
```diff
@@ -70,20 +72,28 @@ export function equityCompensationIssuanceDataToDaml(
...
-    const normalized = normalizeNumericString(v.amount);
-    return parseFloat(normalized) > 0;
-  });
+  const filteredVestings = (d.vestings ?? [])
+    .map((vesting, index) => ({ index, vesting }))
+    .filter(({ vesting }) => {
+      // normalizeNumericString validates strict decimal format and rejects scientific notation
+      const normalized = normalizeNumericString(vesting.amount);
+      return parseFloat(normalized) > 0;
+    });
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/equityCompensationIssuance/createEquityCompensationIssuance.ts]: Fixed in 5453fea. The three issuance writers now share filterAndMapVestings from OpenCapTable/shared/vesting.ts; it preserves pre-filter indexes, strict numeric normalization, amount > 0 filtering, and entity-specific date paths.

File: test/converters/dateBoundaryValidation.test.ts [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+    name: `equity compensation issuance ${field}`,
+    field,
+    fieldPath: `equityCompensationIssuance.${field}`,
+    convert: (value: unknown) =>
+      equityCompensationIssuanceDataToDaml({
+        ...EQUITY_COMPENSATION_WRITE_BASE,
+        [field]: value,
+      }),
+  })),
+];
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply test/converters/dateBoundaryValidation.test.ts:242]: Fixed in 5453fea. The matrix now covers StockClass, StockIssuance, and PlanSecurity approval dates on writes and their real canonical read paths; PlanSecurity reads use the shared EquityCompensation reader because no separate reader exists.

File: src/functions/OpenCapTable/stockClass/getStockClassAsOcf.ts [hunk 10]
```diff
@@ -2,24 +2,103 @@ import type { LedgerJsonApiClient } from '@fairmint/canton-node-sdk';
...
 
+function firstLossyGeneratedPath(
+  source: unknown,
+  encoded: unknown,
+  path: string,
+  ancestors = new WeakSet<object>()
+): string | undefined {
+  if (source === null || typeof source !== 'object') {
+    return Object.is(source, encoded) ? undefined : path;
+  }
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/stockClass/getStockClassAsOcf.ts:26]: Verified against the exact v34 generated codec and JSON Ledger API contract. Returned DAML `Numeric` values are canonical strings; JSON numbers are accepted only on command input. `@daml/types@3.5.0` decodes both `Numeric` and `Int` with `jtv.string()` and encodes them with identity functions—it does not coerce a number to a string. A required numeric primitive is rejected before this comparison. More importantly, an optional numeric primitive is masked to `null` by the generated `withDefault` decoder, and this strict comparison is what prevents that value from being silently discarded. Canonical string values decode/re-encode unchanged, and captured ledger response fixtures use strings. Keeping the guard as-is; loosening primitive equality would reintroduce real data loss.

File: src/functions/OpenCapTable/equityCompensationIssuance/getEquityCompensationIssuanceAsOcf.ts [hunk 11]
```diff
@@ -50,14 +56,12 @@ const twMapPeriodType: Partial<Record<string, PeriodType>> = {
  * Used by both getEquityCompensationIssuanceAsOcf and the damlToOcf dispatcher.
  */
 export function damlEquityCompensationIssuanceDataToNative(d: Record<string, unknown>): OcfEquityCompensationIssuance {
-  const exercise_price = damlMonetaryToNativeWithValidation(
-    d.exercise_price as Record<string, unknown> | null | undefined
-  );
-  const base_price = damlMonetaryToNativeWithValidation(d.base_price as Record<string, unknown> | null | undefined);
+  const exercise_price = damlMonetaryToNativeWithValidation(d.exercise_price);
+  const base_price = damlMonetaryToNativeWithValidation(d.base_price);
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/equityCompensationIssuance/getEquityCompensationIssuanceAsOcf.ts]: Fixed in `02bb2a1`: both equity-compensation monetary reads now pass exact contextual field paths. Added regressions for malformed `exercise_price` and `base_price` payloads.

File: src/functions/OpenCapTable/stockClass/getStockClassAsOcf.ts [hunk 12]
```diff
@@ -111,117 +199,161 @@ export function damlStockClassDataToNative(
...
-                : null;
-          if (tag === 'OcfRoundingCeiling') return 'CEILING';
-          if (tag === 'OcfRoundingFloor') return 'FLOOR';
-          if (tag === 'OcfRoundingNormal') return 'NORMAL';
-          return 'NORMAL';
-        };
-        const roundingType = extractRoundingType(rec.rounding_type);
+        const trigger = right.conversion_trigger;
+        const triggerPath = `${path}.conversion_trigger`;
+        const expectedTriggerId = `ocp-sdk:stock-class:${damlData.id}:conversion-right:${index}:unspecified`;
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/stockClass/getStockClassAsOcf.ts]: Fixed in `02bb2a1`: the storage-only description and deterministic trigger-ID builder now live in `stockClassConversionStorage.ts` and are shared by both writer and reader.

File: src/functions/OpenCapTable/stockClass/stockClassDataToDaml.ts [hunk 13]
```diff
@@ -1,155 +1,130 @@
...
+    percent_of_capitalization: null,
+    ratio: {
+      numerator: normalizeNumericString(right.conversion_mechanism.ratio.numerator),
+      denominator: normalizeNumericString(right.conversion_mechanism.ratio.denominator),
+    },
+    reference_share_price: null,
+    reference_valuation_price_per_share: null,
+    valuation_cap: null,
+  };
+}
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/stockClass/stockClassDataToDaml.ts:179]: Fixed in `02bb2a1`: stock-class writes validate the exact ratio mechanism discriminator, rounding, conversion price, ratio object, and both ratio parts before dereferencing. Ten missing/malformed cases assert structured `OcpValidationError` codes and paths, with no raw `TypeError`.

File: test/schemaAlignment/dateBoundaryInvariants.test.ts [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+          const expectedArgumentCount = node.expression.text === 'triggerFieldsToDaml' ? 2 : 3;
+          if (node.arguments.length !== expectedArgumentCount) {
+            violations.push(
+              `${location(sourceFile, node)} ${node.expression.text} must use its discriminator-correlated signature`
+            );
+          }
+          const fieldPath = node.arguments[expectedArgumentCount - 1];
+          if (fieldPath.getText(sourceFile).includes('[]')) {
+            violations.push(`${location(sourceFile, fieldPath)} trigger array fieldPath must include its index`);
+          }
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply test/schemaAlignment/dateBoundaryInvariants.test.ts:212]: Fixed in `02bb2a1`: the invariant checks the argument count before reading the field-path argument, so signature violations remain actionable instead of crashing the test.

File: src/functions/OpenCapTable/stockClass/getStockClassAsOcf.ts [hunk 15]
```diff
@@ -111,117 +203,161 @@ export function damlStockClassDataToNative(
...
+        if (right.conversion_mechanism !== 'OcfConversionMechanismRatioConversion') {
+          throw new OcpParseError(`Unknown stock class conversion mechanism: ${right.conversion_mechanism}`, {
+            source: `${path}.conversion_mechanism`,
+            code: OcpErrorCodes.SCHEMA_MISMATCH,
+          });
+        }
+        if (!right.ratio) {
+          throw new OcpValidationError(`${path}.conversion_mechanism.ratio`, 'Required OCF ratio is missing', {
+            code: OcpErrorCodes.SCHEMA_MISMATCH,
+          });
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/stockClass/getStockClassAsOcf.ts:231]: Verified against the current read boundary: this path is intentionally canonical OCF, not the private flattened DAML-v34 record. The returned public right is `conversion_mechanism.ratio` (and `conversion_mechanism.conversion_price`), while the sibling DAML fields are a storage adapter detail this PR deliberately hides. Reporting the raw sibling path here would leak that private mismatch and make read-side diagnostics disagree with the public SDK/write boundary. No code change is appropriate.

File: src/functions/OpenCapTable/shared/vesting.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+/** Validate every vesting row, then filter zero-value placeholders while retaining original indexes. */
+export function filterAndMapVestingsToDaml(
+  vestings: readonly VestingInput[] | null | undefined,
+  basePath: string
+): DamlVesting[] {
+  return (vestings ?? [])
+    .map((vesting, index) => {
+      const amountPath = `${basePath}[${index}].amount`;
+      const date = dateStringToDAMLTime(vesting.date, `${basePath}[${index}].date`);
+      const amount = normalizeNumericString(vesting.amount, amountPath);
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/shared/vesting.ts:32]: Fixed in `9d459b2`: `filterAndMapVestingsToDaml` now narrows every row with `isRecord` before dereferencing it and emits an indexed `OcpValidationError` (`INVALID_TYPE`, expected `object`) for null, arrays, and primitives. The shared write-boundary suite covers all three malformed shapes and the existing valid canonicalization path.

File: src/functions/OpenCapTable/convertibleIssuance/createConvertibleIssuance.ts [hunk 17]
```diff
@@ -181,20 +182,37 @@ function mechanismInputToDamlEnum(
...
+                }
+
+                return {
+                  rate: ir?.rate != null ? normalizeNumericString(String(ir.rate), `${interestRatePath}.rate`) : null,
+                  accrual_start_date: dateStringToDAMLTime(accrualStartDate, `${interestRatePath}.accrual_start_date`),
+                  accrual_end_date: optionalDateStringToDAMLTime(
+                    ir?.accrual_end_date,
+                    `${interestRatePath}.accrual_end_date`
+                  ),
+                };
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/convertibleIssuance/createConvertibleIssuance.ts:228]: Fixed in `9d459b2`: convertible-note `interest_rates[i]` now must be a record before any field access and reports the same indexed shape diagnostic as the read converter. Regression coverage exercises null, arrays, and primitives; `rate` is also explicitly narrowed to string/number before numeric normalization.
